File: BlotoutAnalytics/Extention/UIViewController+Extensions.m
//
//  UIViewController+ExtensionsViewController.m
//  BlotoutAnalytics
//
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "UIViewController+Extensions.h"
#import <objc/runtime.h>
#import "BOSharedManager.h"
#import "BOFLogs.h"
#import "BOAUtilities.h"
#import "BOANetworkConstants.h"
#import "BOACaptureModel.h"
#import "BlotoutAnalytics_Internal.h"
#import "BOEventsOperationExecutor.h"

void loadAsUIViewControllerBOFoundationCat(void) {}

@implementation UIViewController (Extensions)

+ (void)load {
  static dispatch_once_t once_token;
  dispatch_once(&once_token,  ^{
    SEL viewDidAppearSelector = @selector(viewDidAppear:);
    SEL viewDidAppearLoggerSelector = @selector(logged_viewDidAppear:);
    Method originalMethod = class_getInstanceMethod(self, viewDidAppearSelector);
    Method extendedMethod = class_getInstanceMethod(self, viewDidAppearLoggerSelector);
    method_exchangeImplementations(originalMethod, extendedMethod);
    
    SEL viewWillDisappearSelector = @selector(viewWillDisappear:);
    SEL viewWillDisappearLoggerSelector = @selector(logged_viewWillDisappear:);
    Method originalDisappearMethod = class_getInstanceMethod(self, viewWillDisappearSelector);
    Method extendedDisappearMethod = class_getInstanceMethod(self, viewWillDisappearLoggerSelector);
    method_exchangeImplementations(originalDisappearMethod, extendedDisappearMethod);
  });
}

+ (UIViewController *)getRootViewControllerFromView:(UIView *)view {
  
  UIViewController *root = view.window.rootViewController;
  return [self topViewController:root];
  
}

+ (UIViewController *)topViewController:(UIViewController *)rootViewController
{
    UIViewController *nextRootViewController = [self nextRootViewController:rootViewController];
    if (nextRootViewController) {
        return [self topViewController:nextRootViewController];
    }

    return rootViewController;
}

+ (UIViewController *)nextRootViewController:(UIViewController *)rootViewController {
    UIViewController *presentedViewController = rootViewController.presentedViewController;
    if (presentedViewController != nil) {
        return presentedViewController;
    }
    if ([rootViewController isKindOfClass:[UINavigationController class]]) {
        UIViewController *lastViewController = ((UINavigationController *)rootViewController).viewControllers.lastObject;
        return lastViewController;
    }
    if ([rootViewController isKindOfClass:[UITabBarController class]]) {
        __auto_type *currentTabViewController = ((UITabBarController*)rootViewController).selectedViewController;
        if (currentTabViewController != nil) {
            return currentTabViewController;
        }
    }
    return nil;
}

-(NSString*)getScreenName:(UIViewController *)viewController {
    NSString *name = [[[viewController class] description] stringByReplacingOccurrencesOfString:@"ViewController" withString:@""];
    if (!name || name.length == 0) {
        name = [viewController title];
        if (name.length == 0) {
            name = @"Unknown";
        }
    }
    return name;
}

- (void)logged_viewWillDisappear:(BOOL)animated {
  @try {
    
    UIViewController *top = [[self class] getRootViewControllerFromView:self.view];
    
    if (!top) {
      return;
    }
    
    [self logged_viewWillDisappear:animated];
    [BOSharedManager sharedInstance].isViewDidAppeared = NO;
    
    if ([BlotoutAnalytics sharedInstance].eventManager == nil) {
      return;
    }
    
    NSString *screen = [self getScreenName:top];
    
    [[BOEventsOperationExecutor sharedInstance] dispatchEventsInBackground:^{
      BOACaptureModel *model = [[BOACaptureModel alloc] initWithEvent:BO_VISIBILITY_HIDDEN properties:nil eventCode:@(BO_EVENT_VISIBILITY_HIDDEN) screenName:screen withType:BO_SYSTEM];
      [[BlotoutAnalytics sharedInstance].eventManager capture:model];
    }];
    
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
}

- (void)logged_viewDidAppear:(BOOL)animated {
  @try {
    
    UIViewController *top = [[self class] getRootViewControllerFromView:self.view];
    
    if (!top) {
      return;
    }
    
    [self logged_viewDidAppear:animated];
    
    if ([BOSharedManager sharedInstance].isViewDidAppeared) {
      return;
    }
    
    [BOSharedManager sharedInstance].isViewDidAppeared = YES;
    
    if ([BlotoutAnalytics sharedInstance].eventManager == nil) {
      return;
    }
    
    NSString *screenName = [self getScreenName:top];
    [BOSharedManager sharedInstance].currentScreenName = screenName;
   
    [[BOEventsOperationExecutor sharedInstance] dispatchEventsInBackground:^{
      BOACaptureModel *model = [[BOACaptureModel alloc] initWithEvent:BO_SDK_START properties:nil eventCode:@(BO_EVENT_SDK_START) screenName:screenName withType:BO_SYSTEM];
      [[BlotoutAnalytics sharedInstance].eventManager capture:model];
    }];
    
  } @catch (NSException *exception) {
      BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
}
@end

File: BlotoutAnalytics/Extention/UIViewController+Extensions.swift
//
//  UIViewController+Extensions.swift
//  BlotoutAnalytics
//
//  Created by Poonam Tiwari on 08/06/22.
//

import Foundation
import UIKit

extension UIViewController {
    
//    @objc func newViewWillAppear(_ animated: Bool) {
//        self.newViewWillAppear(animated) //Incase we need to override this method
//        let viewControllerName = String(describing: type(of: self)).replacingOccurrences(of: "ViewController", with: "", options: .literal, range: nil)
//        print("Swizzled viewWillAppear", viewControllerName)
//    }
//
//    static func swizzleViewWillAppear() {
//        //Make sure This isn't a subclass of UIViewController, So that It applies to all UIViewController childs
//        if self != UIViewController.self {
//            return
//        }
//        let _: () = {
//            let originalSelector = #selector(UIViewController.viewDidAppear(_:))
//            let swizzledSelector = #selector(UIViewController.logged_viewDidAppear(_:))
//            let originalMethod = class_getInstanceMethod(self, originalSelector)
//            let swizzledMethod = class_getInstanceMethod(self, swizzledSelector)
//            method_exchangeImplementations(originalMethod!, swizzledMethod!);
//        }()
//    }
    
    
   static func swizzleVCLoggingMethods()
    {
        let _: () = {
            let originalSelector = #selector(UIViewController.viewDidAppear(_:))
            let swizzledSelector = #selector(UIViewController.logged_viewDidAppear(_:))
            let originalMethod = class_getInstanceMethod(self, originalSelector)
            let swizzledMethod = class_getInstanceMethod(self, swizzledSelector)
            method_exchangeImplementations(originalMethod!, swizzledMethod!);

            
            let originalSelector2 = #selector(UIViewController.viewWillDisappear(_:))
            let swizzledSelector2 = #selector(UIViewController.logged_viewWillDisappear(_:))
            let originalMethod2 = class_getInstanceMethod(self, originalSelector2)
            let swizzledMethod2 = class_getInstanceMethod(self, swizzledSelector2)
            method_exchangeImplementations(originalMethod2!, swizzledMethod2!);
        }()
    }
    
    class func getRootViewController(from view: UIView?) -> UIViewController? {

        let root = view?.window?.rootViewController
        return self.topViewController(root)

    }
    
    class func topViewController(_ rootViewController: UIViewController?) -> UIViewController? {
        let nextRootViewController = self.nextRootViewController(rootViewController)
        if let nextRootViewController = nextRootViewController {
            return self.topViewController(nextRootViewController)
        }

        return rootViewController
    }
    
    class func nextRootViewController(_ rootViewController: UIViewController?) -> UIViewController? {
        let presentedViewController = rootViewController?.presentedViewController
        if let presentedViewController = presentedViewController {
            return presentedViewController
        }
        if rootViewController is UINavigationController {
            let lastViewController = (rootViewController as? UINavigationController)?.viewControllers.last
            return lastViewController
        }
        if rootViewController is UITabBarController {
            let currentTabViewController = (rootViewController as? UITabBarController)?.selectedViewController
            if let currentTabViewController = currentTabViewController {
                return currentTabViewController
            }
        }
        return nil
    }
    
    func getScreenName(_ viewController: UIViewController) -> String? {
        var name = type(of: viewController).typeName
        if  name.count == 0 {
            name = viewController.title ?? ""
            if name.count == 0 {
                name = "Unknown"
            }
        }
        return name
    }

    
    @objc func logged_viewWillDisappear(_ animated: Bool) {

            let top = UIViewController.getRootViewController(from: view)
            
            if top == nil {
                return
            }
            
        //TODO: understand why same method is recursively called???
        
            logged_viewWillDisappear(animated)
        
        //TODO: viewDidAppeared logic fails in case of present, so commenting, lets change if we see the usecase for this
           // BOSharedManager.sharedInstance.isViewDidAppeared = false
            
            if BlotoutAnalytics.sharedInstance.eventManager == nil {
                return
            }
            
            let screen = getScreenName(top!)
            BOEventsOperationExecutor.sharedInstance.dispatchEvents(inBackground: {
                let model = BOACaptureModel(event: BO_VISIBILITY_HIDDEN, properties: nil, screenName: screen, withType: BO_SYSTEM)
                BlotoutAnalytics.sharedInstance.eventManager.capture(model)
            })
    }
    
    @objc func logged_viewDidAppear(_ animated: Bool) {

            let top = UIViewController.getRootViewController(from: view)

            if top == nil {
                return
            }

        //TODO: understand why same method is recursively called???
            logged_viewDidAppear(animated)

        //TODO: viewDidAppeared logic fails in case of present, so commenting, lets change if we see the usecase for this
//            if BOSharedManager.sharedInstance.isViewDidAppeared {
//                return
//            }
          //  BOSharedManager.sharedInstance.isViewDidAppeared = true

            if BlotoutAnalytics.sharedInstance.eventManager == nil {
                return
            }
            let screenName = getScreenName(top!)
            BOSharedManager.sharedInstance.currentScreenName = screenName

            BOEventsOperationExecutor.sharedInstance.dispatchEvents(inBackground: {
                let model = BOACaptureModel(event: BO_SDK_START, properties: nil, screenName: screenName, withType: BO_SYSTEM)
                BlotoutAnalytics.sharedInstance.eventManager.capture(model)
            })
        
    }
}

// Bridge to Obj-C
extension NSObject {
    class var typeName: String {
        let type = String(describing: self)
        return type
    }
}
